site/docs/blog/posts/2024-05-new-cli.md

```diff
@@ -0,0 +1,119 @@
+a command line. It exists since the early days of Nessie.  Since then,
+`pynessie` did not change very much and is still using the Nessie REST
+API v1, which is deprecated and will be removed at some point.
+`pynessie` also does not have support for the more complex OAuth2
+device-code and authorization-code flows. 
+
+Updating `pynessie` to work with Nessie REST API v2 basically requires
+to re-implement a lot of code, starting with JSON de/serialization,
+model objects, etc.
+
+We were thinking about how to deal with a CLI tool for Nessie,
+considering a bunch of options and choices:
+
+## Shall we continue with Python or use Java?
+
+You may say: Python, that's it! Much easier to run! With _easier_
+people usually think of "just running it" from bash or zsh. But is
+it really easier run? You have to setup a virtual-env and have the
+(right) Python version in place.
+
+Dependency management in Python is (at least from our perspective)
+not an easy thing. API compatibility between versions of the
+dependencies is not always maintained.
+
+TL;DR we think, there's not really a benefit from using Python.
+
+## Feature "parity"
+
+`pynessie`, since written in Python, requires that we duplicate
+everything (Nessie REST API model classes, REST calls, 
+authentication mechanisms, also OAuth2 flows) - resulting in a
+duplicate engineering effort. It's more work to do all that in
+Python.
+
+## REPL?
+
+Having a REPL for the Nessie CLI, including a "scripting language",
+would be cool, no? pynessie does not have that.
+
+## Tool startup time
+
+Admittedly, pynessie starts pretty quickly. However, nowadays
+Java tools start up pretty quick as well - if done right, you'll
+not notice, especially for an interactive tool (think: REPL).
+
+## What about the `nessie-content-generator` tool?
+
+That tool was, as the name suggests, a tool to generate dummy
+content in Nessie for testing purposes. It's never really been
+meant to be used as a CLI (replacement).
+
+`nessie-content-generator` is relatively easy to use - so there
+is some risk that you generate a lot of "broken" content in a
+production Nessie repository, maybe not the best thing.
+
+## Java it is!
+
+We decided to go with Java, we have all the libraries that we need:
+
+* picocli is really neat for command line options et al
+* jline + jansi is a really cool library combination when working
+  with terminals - colors, REPL tool functionality, command history,
+  etc etc
+* Nessie client - we already have a fully featured client for Java
+  that also already supports all the OAuth2 flows, including
+  device-code and authorization-code.
+
+## Do we want Quarkus to run the CLI/REPL?
+
+Sure, Quarkus offers a lot of things, but there's nothing we really
+need - it can't even "eliminate" dead code for us, because nothing
+in the Nessie Java client uses CDI.
+
+Also, with Quarkus we'd have to require Java 17, but requiring "just"
+Java 11 is probably nicer. So we don't use Quarkus.
```

Comment:
Is it? This is an end product, not a library. And a client-side one. The only thing the user needs to run it, is a java runtime installed. So why not 17 or even 21?

Reply:
I suspect data lake people to have rather 11 than 17 or 21